sn_networking/src/cmd.rs

```diff
@@ -980,26 +984,30 @@ impl SwarmDriver {
         let acceptable_distance_range = self.get_request_range();
         let target_key = target_address.as_kbucket_key();
 
-        let peers = self
+        let sorted_peers: Vec<_> = self
             .swarm
             .behaviour_mut()
             .kademlia
             .get_closest_local_peers(&target_key)
-            .filter_map(|key| {
-                // here we compare _bucket_, not the exact distance.
-                // We want to include peers that are just outside the range
-                // Such that we can and will exceed the range in a search eventually
-                if acceptable_distance_range.ilog2() < target_key.distance(&key).ilog2() {
-                    return None;
-                }
+            .collect();
 
-                // Map KBucketKey<PeerId> to PeerId.
-                let peer_id = key.into_preimage();
-                Some(peer_id)
+        // Binary search to find the index where we exceed the acceptable range
+        let split_index = sorted_peers
+            .binary_search_by(|key| {
+                let distance = target_key.distance(key);
+                if distance >= acceptable_distance_range {
```

Comment:
note this is a change from `kbucket index range check` to `exact range check`,
which could be too stict that not leaving any margin for edge cases.